Spec file with multiple flows in the `oauth` schema leads into duplicated `@OauthAuthenticationMarker` in `CompositeAuthenticationProvider`

Hi team,

we are currently trying to upgrade the Quarkus version (to `3.15.2`) in Kogito Runtimes, and after upgrading the [quarkus-openapi-generator](https://github.com/quarkiverse/quarkus-openapi-generator) to `2.7.1-lts`, we noticed that some of our IT tests were failing when the openapi spec files have multiple flows in the `oauth` Security Schema, ej: [openapi3.0-security.yaml](https://github.com/apache/incubator-kie-kogito-runtimes/blob/main/quarkus/integration-tests/integration-tests-quarkus-openapi-client/src/main/resources/specs/openapi3.0-security.yaml)

With this setup, the extension generates the `CompositeAuthenticationProvider` with a duplicated `OauthAuthenticationMarker` annotation, which leads into the `GeneratorProcessor` trying to register twice an `AuthProvider` Synthetic Bean using the same identifier breaking the app build.

* Generated `CompositeAuthenticationProvider`:
```
@Priority(1000)
@AuthenticationMarkers({
 @OauthAuthenticationMarker(name = "oauth", openApiSpecId = "openapi3_0_security_yaml" ),
 @OauthAuthenticationMarker(name = "oauth", openApiSpecId = "openapi3_0_security_yaml" )
})
public class CompositeAuthenticationProvider implements ClientRequestFilter {
...
}
```

* Error logs when building the application

```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:3.15.2:build (default) on project integration-tests-quarkus-openapi-client: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.arc.deployment.SyntheticBeansProcessor#initStatic threw an exception: java.lang.IllegalStateException: A synthetic bean with identifier Wb25h6J20BU0nnvE3Cp26vFuAL8 is already registered: SYNTHETIC bean [types=[io.quarkiverse.openapi.generator.providers.AuthProvider, java.lang.Object], qualifiers=[@Any, @io.quarkiverse.openapi.generator.OpenApiSpec(openApiSpecId = "openapi3_0_security_yaml"), @io.quarkiverse.openapi.generator.AuthName(name = "oauth")], target=n/a]
[ERROR] 	at io.quarkus.arc.processor.BeanDeployment.addSyntheticBean(BeanDeployment.java:1511)
[ERROR] 	at io.quarkus.arc.processor.BeanDeployment$BeanRegistrationContextImpl.accept(BeanDeployment.java:1866)
[ERROR] 	at io.quarkus.arc.processor.BeanDeployment$BeanRegistrationContextImpl.accept(BeanDeployment.java:1835)
[ERROR] 	at io.quarkus.arc.processor.BeanConfigurator.done(BeanConfigurator.java:114)
[ERROR] 	at io.quarkus.arc.deployment.SyntheticBeansProcessor.configureSyntheticBean(SyntheticBeansProcessor.java:92)
[ERROR] 	at io.quarkus.arc.deployment.SyntheticBeansProcessor.initStatic(SyntheticBeansProcessor.java:39)
[ERROR] 	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:732)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:840)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:483)
[ERROR] -> [Help 1]
```

We can adjust our tests to avoid this issue, but I think it is worth it to have it reported.

--- a/client/deployment/src/main/java/io/quarkiverse/openapi/generator/deployment/template/OpenApiNamespaceResolver.java
+++ b/client/deployment/src/main/java/io/quarkiverse/openapi/generator/deployment/template/OpenApiNamespaceResolver.java
@@ -39,6 +39,7 @@ public class OpenApiNamespaceResolver implements NamespaceResolver {
      * @param codegenConfig Map with the model codegen properties
      * @return true if the given model class should generate the deprecated attributes
      */
+    @SuppressWarnings("unused")
     public boolean genDeprecatedModelAttr(final String pkg, final String classname,
             final HashMap<String, Object> codegenConfig) {
         final String key = String.format("%s.%s.%s", pkg, classname, GENERATE_DEPRECATED_PROP);
@@ -51,16 +52,19 @@ public class OpenApiNamespaceResolver implements NamespaceResolver {
      * @param codegenConfig Map with the model codegen properties
      * @return true if the given model class should generate the deprecated attributes
      */
+    @SuppressWarnings("unused")
     public boolean genDeprecatedApiAttr(final String pkg, final String classname,
             final HashMap<String, Object> codegenConfig) {
         final String key = String.format("%s.%s.%s", pkg, classname, GENERATE_DEPRECATED_PROP);
         return Boolean.parseBoolean(codegenConfig.getOrDefault(key, "true").toString());
     }
 
+    @SuppressWarnings("unused")
     public String parseUri(String uri) {
         return OpenApiGeneratorOutputPaths.getRelativePath(Path.of(uri)).toString().replace(File.separatorChar, '/');
     }
 
+    @SuppressWarnings("unused")
     public boolean hasAuthMethods(OperationMap operations) {
         return operations != null && operations.getOperation().stream().anyMatch(operation -> operation.hasAuthMethods);
     }
@@ -75,6 +79,7 @@ public class OpenApiNamespaceResolver implements NamespaceResolver {
      * @see "resources/templates/libraries/microprofile/auth/compositeAuthenticationProvider.qute"
      * @return The list filtered by unique auth name
      */
+    @SuppressWarnings("unused")
     public List<CodegenSecurity> getUniqueOAuthOperations(List<CodegenSecurity> oauthOperations) {
         if (oauthOperations != null) {
             return new ArrayList<>(oauthOperations.stream()

--- a/client/deployment/src/main/java/io/quarkiverse/openapi/generator/deployment/template/OpenApiNamespaceResolver.java
+++ b/client/deployment/src/main/java/io/quarkiverse/openapi/generator/deployment/template/OpenApiNamespaceResolver.java
@@ -1,6 +1,7 @@
 package io.quarkiverse.openapi.generator.deployment.template;
 
 import java.io.File;
+import java.lang.reflect.Method;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Collections;
@@ -74,12 +75,12 @@ public class OpenApiNamespaceResolver implements NamespaceResolver {
      * @see "resources/templates/libraries/microprofile/auth/compositeAuthenticationProvider.qute"
      * @return The list filtered by unique auth name
      */
-    public List<CodegenSecurity> getUniqueOAuthOperations(ArrayList<CodegenSecurity> oauthOperations) {
+    public List<CodegenSecurity> getUniqueOAuthOperations(List<CodegenSecurity> oauthOperations) {
         if (oauthOperations != null) {
-            return oauthOperations.stream()
+            return new ArrayList<>(oauthOperations.stream()
                     .collect(Collectors.toMap(security -> security.name, security -> security,
                             (existing, replacement) -> existing, LinkedHashMap::new))
-                    .values().stream().toList();
+                    .values());
         }
         return Collections.emptyList();
     }
@@ -87,19 +88,49 @@ public class OpenApiNamespaceResolver implements NamespaceResolver {
     @Override
     public CompletionStage<Object> resolve(EvalContext context) {
         try {
-            Class<?>[] classArgs = new Class[context.getParams().size()];
             Object[] args = new Object[context.getParams().size()];
+            Class<?>[] classArgs = new Class[context.getParams().size()];
+
             int i = 0;
             for (Expression expr : context.getParams()) {
                 args[i] = context.evaluate(expr).toCompletableFuture().get();
                 classArgs[i] = args[i].getClass();
                 i++;
             }
-            return CompletableFuture
-                    .completedFuture(this.getClass().getMethod(context.getName(), classArgs).invoke(this, args));
+
+            Method targetMethod = findCompatibleMethod(context.getName(), classArgs);
+            if (targetMethod == null) {
+                throw new NoSuchMethodException("No compatible method found for: " + context.getName());
+            }
+
+            return CompletableFuture.completedFuture(targetMethod.invoke(this, args));
         } catch (ReflectiveOperationException | InterruptedException | ExecutionException ex) {
             return CompletableFuture.failedStage(ex);
         }
+    }
+
+    private Method findCompatibleMethod(String methodName, Class<?>[] argTypes) {
+        for (Method method : this.getClass().getMethods()) {
+            if (method.getName().equals(methodName)) {
+                Class<?>[] paramTypes = method.getParameterTypes();
+                if (isAssignable(paramTypes, argTypes)) {
+                    return method;
+                }
+            }
+        }
+        return null;
+    }
+
+    private boolean isAssignable(Class<?>[] paramTypes, Class<?>[] argTypes) {
+        if (paramTypes.length != argTypes.length) {
+            return false;
+        }
+        for (int i = 0; i < paramTypes.length; i++) {
+            if (!paramTypes[i].isAssignableFrom(argTypes[i])) {
+                return false;
+            }
+        }
+        return true;
     }
 
     @Override

--- a/client/deployment/src/main/java/io/quarkiverse/openapi/generator/deployment/template/OpenApiNamespaceResolver.java
+++ b/client/deployment/src/main/java/io/quarkiverse/openapi/generator/deployment/template/OpenApiNamespaceResolver.java
@@ -2,11 +2,17 @@ package io.quarkiverse.openapi.generator.deployment.template;
 
 import java.io.File;
 import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
 import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
 import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
 
+import org.openapitools.codegen.CodegenSecurity;
 import org.openapitools.codegen.model.OperationMap;
 
 import io.quarkiverse.openapi.generator.deployment.codegen.OpenApiGeneratorOutputPaths;
@@ -56,6 +62,26 @@ public class OpenApiNamespaceResolver implements NamespaceResolver {
 
     public boolean hasAuthMethods(OperationMap operations) {
         return operations != null && operations.getOperation().stream().anyMatch(operation -> operation.hasAuthMethods);
+    }
+
+    /**
+     * Ignore the OAuth flows by filtering every oauth instance by name. The inner openapi-generator library duplicates the
+     * OAuth instances per flow in the openapi spec.
+     * So a specification file with more than one flow defined has two entries in the list. For now, we do not use this
+     * information in runtime so it can be safely filtered and ignored.
+     *
+     * @param oauthOperations passed through the Qute template
+     * @see "resources/templates/libraries/microprofile/auth/compositeAuthenticationProvider.qute"
+     * @return The list filtered by unique auth name
+     */
+    public List<CodegenSecurity> getUniqueOAuthOperations(ArrayList<CodegenSecurity> oauthOperations) {
+        if (oauthOperations != null) {
+            return oauthOperations.stream()
+                    .collect(Collectors.toMap(security -> security.name, security -> security,
+                            (existing, replacement) -> existing, LinkedHashMap::new))
+                    .values().stream().toList();
+        }
+        return Collections.emptyList();
     }
 
     @Override

--- a/client/deployment/src/main/resources/templates/libraries/microprofile/auth/compositeAuthenticationProvider.qute
+++ b/client/deployment/src/main/resources/templates/libraries/microprofile/auth/compositeAuthenticationProvider.qute
@@ -1,9 +1,11 @@
 package {apiPackage}.auth;
 
 @jakarta.annotation.Priority(jakarta.ws.rs.Priorities.AUTHENTICATION)
-{#for auth in oauthMethods.orEmpty}
-@io.quarkiverse.openapi.generator.markers.OauthAuthenticationMarker(name="{auth.name}", openApiSpecId="{configKey}")
+{#if oauthMethods.orEmpty.size > 0}
+{#for auth in openapi:getUniqueOAuthOperations(oauthMethods)}
+@io.quarkiverse.openapi.generator.markers.OauthAuthenticationMarker(name="{auth.name}", openApiSpecId="{quarkus-generator.openApiSpecId}")
 {/for}
+{/if}
 {#for auth in httpBasicMethods.orEmpty}
 @io.quarkiverse.openapi.generator.markers.BasicAuthenticationMarker(name="{auth.name}", openApiSpecId="{quarkus-generator.openApiSpecId}")
 {/for}

--- a/client/deployment/src/main/resources/templates/libraries/microprofile/auth/compositeAuthenticationProvider.qute
+++ b/client/deployment/src/main/resources/templates/libraries/microprofile/auth/compositeAuthenticationProvider.qute
@@ -1,11 +1,9 @@
 package {apiPackage}.auth;
 
 @jakarta.annotation.Priority(jakarta.ws.rs.Priorities.AUTHENTICATION)
-{#if oauthMethods.orEmpty.size > 0}
-{#for auth in openapi:getUniqueOAuthOperations(oauthMethods)}
+{#for auth in openapi:getUniqueOAuthOperations(oauthMethods.orEmpty)}
 @io.quarkiverse.openapi.generator.markers.OauthAuthenticationMarker(name="{auth.name}", openApiSpecId="{quarkus-generator.openApiSpecId}")
 {/for}
-{/if}
 {#for auth in httpBasicMethods.orEmpty}
 @io.quarkiverse.openapi.generator.markers.BasicAuthenticationMarker(name="{auth.name}", openApiSpecId="{quarkus-generator.openApiSpecId}")
 {/for}

--- a/client/deployment/src/test/java/io/quarkiverse/openapi/generator/deployment/wrapper/OpenApiClientGeneratorWrapperTest.java
+++ b/client/deployment/src/test/java/io/quarkiverse/openapi/generator/deployment/wrapper/OpenApiClientGeneratorWrapperTest.java
@@ -33,6 +33,8 @@ import com.github.javaparser.ast.body.MethodDeclaration;
 import com.github.javaparser.ast.body.Parameter;
 import com.github.javaparser.ast.body.VariableDeclarator;
 import com.github.javaparser.ast.expr.AnnotationExpr;
+import com.github.javaparser.ast.expr.Expression;
+import com.github.javaparser.ast.expr.MemberValuePair;
 import com.github.javaparser.ast.expr.SimpleName;
 import com.github.javaparser.ast.expr.SingleMemberAnnotationExpr;
 import com.github.javaparser.ast.nodeTypes.NodeWithName;
@@ -48,6 +50,38 @@ public class OpenApiClientGeneratorWrapperTest {
     private static Optional<MethodDeclaration> getMethodDeclarationByIdentifier(List<MethodDeclaration> methodDeclarations,
             String methodName) {
         return methodDeclarations.stream().filter(md -> md.getName().getIdentifier().equals(methodName)).findAny();
+    }
+
+    @Test
+    void verifyOAuthDuplicateAnnotationOnCompositeAuthProvider() throws URISyntaxException, FileNotFoundException {
+        OpenApiClientGeneratorWrapper generatorWrapper = createGeneratorWrapper("issue-933-security.yaml");
+        final List<File> generatedFiles = generatorWrapper.generate("org.issue933");
+
+        assertNotNull(generatedFiles);
+        assertFalse(generatedFiles.isEmpty());
+
+        final Optional<File> authProviderFile = generatedFiles.stream()
+                .filter(f -> f.getName().endsWith("CompositeAuthenticationProvider.java")).findFirst();
+        assertThat(authProviderFile).isPresent();
+
+        CompilationUnit compilationUnit = StaticJavaParser.parse(authProviderFile.orElseThrow());
+        // Get the class declaration
+        ClassOrInterfaceDeclaration classDeclaration = compilationUnit.findFirst(ClassOrInterfaceDeclaration.class)
+                .orElseThrow(() -> new AssertionError("Class not found in the file"));
+
+        // Collect all OauthAuthenticationMarker annotations
+        long oauthAnnotationsCount = classDeclaration.getAnnotations().stream()
+                .filter(annotation -> annotation.getNameAsString()
+                        .equals("io.quarkiverse.openapi.generator.markers.OauthAuthenticationMarker"))
+                .filter(Expression::isNormalAnnotationExpr)
+                .filter(annotation -> annotation
+                        .findFirst(MemberValuePair.class,
+                                pair -> pair.getNameAsString().equals("name") && pair.getValue().toString().equals("\"oauth\""))
+                        .isPresent())
+                .count();
+
+        // Assert that there's exactly one annotation with name=oauth
+        assertThat(oauthAnnotationsCount).isEqualTo(1);
     }
 
     @Test

--- a/client/deployment/src/test/resources/openapi/issue-933-security.yaml
+++ b/client/deployment/src/test/resources/openapi/issue-933-security.yaml
@@ -1,0 +1,67 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+---
+openapi: 3.0.3
+info:
+  title: Generated API
+  version: "1.0"
+paths:
+  /:
+    post:
+      operationId: doOperation
+      security:
+        - client_id: [ ]
+        - oauth: [ read, write ]
+        - bearerAuth: [ ]
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/MultiplicationOperation'
+      responses:
+        "200":
+          description: OK
+components:
+  schemas:
+    MultiplicationOperation:
+      type: object
+  securitySchemes:
+    client_id:
+      type: apiKey
+      in: header
+      name: X-Client-Id
+      x-key-type: clientId
+    bearerAuth:
+      type: http
+      scheme: bearer
+    oauth:
+      type: oauth2
+      flows:
+        authorizationCode:
+          authorizationUrl: https://example.com/oauth/authorize
+          tokenUrl: https://example.com/oauth/token
+          scopes:
+            read: Grants read access
+            write: Grants write access
+            admin: Grants read and write access to administrative information
+        clientCredentials:
+          tokenUrl: http://localhost:8382/oauth/token
+          scopes:
+            read: read

--- a/client/deployment/src/test/resources/openapi/issue-933-security.yaml
+++ b/client/deployment/src/test/resources/openapi/issue-933-security.yaml
@@ -1,23 +1,3 @@
-#
-# Licensed to the Apache Software Foundation (ASF) under one
-# or more contributor license agreements.  See the NOTICE file
-# distributed with this work for additional information
-# regarding copyright ownership.  The ASF licenses this file
-# to you under the Apache License, Version 2.0 (the
-# "License"); you may not use this file except in compliance
-# with the License.  You may obtain a copy of the License at
-#
-#   http://www.apache.org/licenses/LICENSE-2.0
-#
-# Unless required by applicable law or agreed to in writing,
-# software distributed under the License is distributed on an
-# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
-# KIND, either express or implied.  See the License for the
-# specific language governing permissions and limitations
-# under the License.
-#
-
----
 openapi: 3.0.3
 info:
   title: Generated API

